File: snippets/fsharp/System/Exception/InnerException/innerex.fs
module Example

//<Snippet1>
open System

type AppException =
    inherit Exception
    
    new (message: string) = { inherit Exception(message) }

    new (message: string, inner) = { inherit Exception(message, inner) }

let throwInner () =
    raise (AppException "Exception in throwInner function.")
    ()

let catchInner () =
    try
        throwInner ()
    with :? AppException as e ->
        raise (AppException("Error in catchInner caused by calling the throwInner function.", e) )

[<EntryPoint>]
let main _ =
    try
        catchInner ()
    with :? AppException as e ->
        printfn "In with block of main function."
        printfn $"Caught: {e.Message}"
        if e.InnerException <> null then
            printfn $"Inner exception: {e.InnerException}"
    0
// The example displays the following output:
//    In with block of main function.
//    Caught: Error in catchInner caused by calling the throwInner function.
//    Inner exception: Example+AppException: Exception in throwInner function.
//       at Example.throwInner()
//       at Example.catchInner()
// </Snippet1>

File: snippets/csharp/System/Exception/InnerException/innerex.cs
//<Snippet1>
using System;

public class AppException : Exception
{
    public AppException(string message) : base(message)
    { }

    public AppException(string message, Exception inner) : base(message, inner) { }
}

public class Example
{
    public static void Main()
    {
        Example ex = new();

        try
        {
            ex.CatchInner();
        }
        catch (AppException e)
        {
            Console.WriteLine("In catch block of Main method.");
            Console.WriteLine($"Caught: {e.Message}");
            if (e.InnerException != null)
                Console.WriteLine($"Inner exception: {e.InnerException}");
        }
    }

    public void ThrowInner() => throw new AppException("Exception in ThrowInner method.");

    public void CatchInner()
    {
        try
        {
            this.ThrowInner();
        }
        catch (AppException e)
        {
            throw new AppException("Error in CatchInner caused by calling the ThrowInner method.", e);
        }
    }
}
// The example displays the following output:
//    In catch block of Main method.
//    Caught: Error in CatchInner caused by calling the ThrowInner method.
//    Inner exception: AppException: Exception in ThrowInner method.
//       at Example.ThrowInner()
//       at Example.CatchInner()
// </Snippet1>
